mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -289,17 +301,28 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
-
-    val sentences: RDD[Array[Int]] = words.mapPartitions { iter =>
+    // each partition is a collection of sentences, will be translated into arrays of Index integer
+    val sentences: RDD[Array[Int]] = dataset.mapPartitions { sentenceIter =>
       new Iterator[Array[Int]] {
-        def hasNext: Boolean = iter.hasNext
+        var wordIter: Iterator[String] = null
+
+        def hasNext: Boolean = sentenceIter.hasNext || (wordIter != null && wordIter.hasNext)
 
         def next(): Array[Int] = {
           val sentence = ArrayBuilder.make[Int]
           var sentenceLength = 0
-          while (iter.hasNext && sentenceLength < MAX_SENTENCE_LENGTH) {
-            val word = bcVocabHash.value.get(iter.next())
-            word match {
+          // do translation of each word into its index in the vocabulary,
```

Comment:
sorry, I can't do much about spark-perf thing, I even still didn't get time to figure out what's blocking me from running the whole test locally.
I(and my compiler) also don't aware of the existence of flatMapPartitions function, but I do made a version suppose to do whatever Sean suggested. Please review and help to do the perf-test for two latest versions.
BTW: I don't worry much about the perf difference since it shouldn't be much, even minutes(not quite possible) of difference for each partition, just mean possibly tens minutes of overall penalty, shouldn't matter much for a job runs hours. Of course, seeing the data is more convincing. My point is that, If diff is minor, we'd better optimize for readability.